File: testsuites/HLT/Benchmark/concurrency-benchmark/PerThreadMemUsage/go/PerThreadMemUsage.go
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2026. All rights reserved.
 * This source file is part of the Cangjie project, licensed under Apache-2.0
 * with Runtime Library Exception.
 * 
 * See https://cangjie-lang.cn/pages/LICENSE for license information.
 */

package main
import (
    _ "fmt"
    "sync"
    "os"
    "strconv"
    "sync/atomic"
)

var allReady int32
var mutex sync.Mutex

func create(cnt *int32) {
  go func() {
    x := atomic.AddInt32(cnt, -1)
    if x == 0 {
      atomic.StoreInt32(&allReady, 1)
    }

    mutex.Lock()
  }()
}


func measure(count int) {
  var cnt int32
  cnt = int32(count)

  mutex.Lock()
  for i := 0; i < count; i++ {
    create(&cnt)
  }

  for atomic.LoadInt32(&allReady) == 0 { }

  println("ready")

  for true { }
}

func main() {
  args := os.Args
  count, _ := strconv.Atoi(args[1])

  measure(count)
}

File: testsuites/HLT/Benchmark/concurrency-benchmark/PerThreadMemUsage/loom/PerThreadMemUsage.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2026. All rights reserved.
 * This source file is part of the Cangjie project, licensed under Apache-2.0
 * with Runtime Library Exception.
 * 
 * See https://cangjie-lang.cn/pages/LICENSE for license information.
 */


import java.util.concurrent.*;
import java.util.concurrent.atomic.*;
import java.util.concurrent.locks.*;

public class PerThreadMemUsage {

    public static volatile boolean allReady = false;
    public static final ReentrantLock m = new ReentrantLock();

    public static void create(AtomicInteger cnt) {
        Thread.startVirtualThread(() -> {
            int x = cnt.getAndAdd(-1) - 1;
            if (x == 0) {
                allReady = true;
            }
            m.lock();
        });
    }

    static void measure(int count) throws InterruptedException, ExecutionException {
        AtomicInteger cnt = new AtomicInteger(count);

        m.lock();
        for (int i = 0; i < count; ++i) {
            create(cnt);
        }

        while (!allReady) { }

        System.out.println("ready");

        while (true) { }
    }


    public static void main(String[] args) throws InterruptedException, ExecutionException {
        int count = Integer.parseInt(args[0]);

        measure(count);
    }
}
